azavea/raster-foundry
# Add db model/endpoints for geospatial tools

## Overview

This commit adds database models and API endpoints for user-created geo-processing tools.

TODO: Filtering is not implemented in this commit, another task will need to be added for it.

### Checklist


## Testing Instructions

 * Start development VM once on this branch
 * If you have not done so, load development data (`./scripts/load_development_data`) and run migrations (`mg update` and `mg apply` once in an `sbt` shell)

 * Open a scala shell
   ```scala
    project app
    console
   ```
 * Verify that the following snippet creates/deletes/retrieves tools successfully

```scala
// testing - increment testId to test subsequent times
val testId = 32
val userId = "auth0|57b744b09da7f9936f14fd5b"

// create two tool tags
val toolTag1Create = ToolTag.Create(publicOrgId, s"Test tag1 $testId")
val toolTag1CreatedFuture = ToolTags.insertToolTag(toolTag1Create, userId)
val toolTag1Created = Await.result(toolTag1CreatedFuture, 10 seconds)
val toolTag1RetrievedFuture = ToolTags.getToolTag(toolTag1Created.id)
val toolTag1Retrieved = Await.result(toolTag1RetrievedFuture, 10 seconds)
val toolTag2Create = ToolTag.Create(publicOrgId, s"Test tag2 $testId")
val toolTag2CreatedFuture = ToolTags.insertToolTag(toolTag2Create, userId)
val toolTag2Created = Await.result(toolTag2CreatedFuture, 10 seconds)
val toolTag2RetrievedFuture = ToolTags.getToolTag(toolTag2Created.id)
val toolTag2Retrieved = Await.result(toolTag2RetrievedFuture, 10 seconds)

// create two tool categories
val toolCategory1Create = ToolCategory.Create(s"Test category1 $testId")
val toolCategory1CreatedFuture = ToolCategories.insertToolCategory(toolCategory1Create, userId)
val toolCategory1Created = Await.result(toolCategory1CreatedFuture, 10 seconds)
val toolCategory1RetrievedFuture = ToolCategories.getToolCategory(toolCategory1Created.id)
val toolCategory1Retrieved = Await.result(toolCategory1RetrievedFuture, 10 seconds)
val toolCategory2Create = ToolCategory.Create(s"Test category2 $testId")
val toolCategory2CreatedFuture = ToolCategories.insertToolCategory(toolCategory2Create, userId)
val toolCategory2Created = Await.result(toolCategory2CreatedFuture, 10 seconds)
val toolCategory2RetrievedFuture = ToolCategories.getToolCategory(toolCategory2Created.id)
val toolCategory2Retrieved = Await.result(toolCategory2RetrievedFuture, 10 seconds)

// create tool with one tag, one category
val tool1Create = Tool.Create(publicOrgId, s"1tag1category$testId", s"desc$testId", s"req$testId", s"lic$testId",
    Visibility.Public, List(s"ds$testId"), 1.1f, List(toolTag1Retrieved.get.id), List(toolCategory1Retrieved.get.id))
val tool1CreatedFuture = Tools.insertTool(tool1Create, userId)
val tool1Created = Await.result(tool1CreatedFuture, 10 seconds)
val tool1RetrievedFuture = Tools.getTool(tool1Created.id)
val tool1Retrieved = Await.result(tool1RetrievedFuture, 10 seconds)
// Should print one tag, one category
println(s"tags: ${tool1Retrieved.get.tags}, categories: ${tool1Retrieved.get.categories}")

// create tool with one tag, zero categories
val tool2Create = Tool.Create(publicOrgId, s"1tag0category$testId", s"desc$testId", s"req$testId", s"lic$testId",
    Visibility.Public, List(s"ds$testId"), 1.1f, List(toolTag1Retrieved.get.id), List())
val tool2CreatedFuture = Tools.insertTool(tool2Create, userId)
val tool2Created = Await.result(tool2CreatedFuture, 10 seconds)
val tool2RetrievedFuture = Tools.getTool(tool2Created.id)
val tool2Retrieved = Await.result(tool2RetrievedFuture, 10 seconds)
// Should print one tag, zero categories
println(s"tags: ${tool2Retrieved.get.tags}, categories: ${tool2Retrieved.get.categories}")

// create tool with zero tags, one category
val tool3Create = Tool.Create(publicOrgId, s"0tag1category$testId", s"desc$testId", s"req$testId", s"lic$testId",
    Visibility.Public, List(s"ds$testId"), 1.1f, List(), List(toolCategory1Retrieved.get.id))
val tool3CreatedFuture = Tools.insertTool(tool3Create, userId)
val tool3Created = Await.result(tool3CreatedFuture, 10 seconds)
val tool3RetrievedFuture = Tools.getTool(tool3Created.id)
val tool3Retrieved = Await.result(tool3RetrievedFuture, 10 seconds)
// Should print zero tags, one category
println(s"tags: ${tool3Retrieved.get.tags}, categories: ${tool3Retrieved.get.categories}")

// create tool with two tags, two categories
val tool4Create = Tool.Create(publicOrgId, s"2tag2category$testId", s"desc$testId", s"req$testId", s"lic$testId",
    Visibility.Public, List(s"ds$testId"), 1.1f, List(toolTag1Retrieved.get.id, toolTag2Retrieved.get.id),
    List(toolCategory1Retrieved.get.id, toolCategory2Retrieved.get.id))
val tool4CreatedFuture = Tools.insertTool(tool4Create, userId)
val tool4Created = Await.result(tool4CreatedFuture, 10 seconds)
val tool4RetrievedFuture = Tools.getTool(tool4Created.id)
val tool4Retrieved = Await.result(tool4RetrievedFuture, 10 seconds)
// Should print two tags, two categories
println(s"tags: ${tool4Retrieved.get.tags}, categories: ${tool4Retrieved.get.categories}")

// delete one of the tools
val tool3DeletedFuture = Tools.deleteTool(tool3Created.id)
val tool3Deleted = Await.result(tool3DeletedFuture, 10 seconds)
// Should be a return of 1 (deleted)

// try deleting again to make sure it's not there
val tool3DeletedFuture = Tools.deleteTool(tool3Created.id)
val tool3Deleted = Await.result(tool3DeletedFuture, 10 seconds)
// Should be a return of 0 (no record)
```

 * Start API: `./scripts/console app-server "./sbt app/run"`
 * Test that a `401` status is returned for un-authenticated requests:
```bash
curl -H "Content-Type: application/json" \
     http://localhost:9000/api/tools/

 * Test that authenticated requests succeed:
```bash
curl -H "Content-Type: application/json" \
     -H "Authorization: Bearer <TOKEN>" \
     http://localhost:9000/api/tools/
```
* Test that tools can be created:
```bash
curl -H "Content-Type: application/json" \
     -X POST \
     -d '{"organizationId": "dfac6307-b5ef-43f7-beda-b9f208bb7726", "stars": 3.1, "description": "Test description", "title": "Test tool from API 8", "visibility": "PUBLIC", "requirements": "tr", "license": "tl", "compatibleDataSources": ["ds1"], "tags": [], "categories": [] }' \
     -H "Authorization: Bearer <TOKEN>" \
     http://localhost:9000/api/tools/
```
* Test that newly created tool is returned (plugging in id returned from previous step):
```
curl -H "Content-Type: application/json" \
     -H "Authorization: Bearer <TOKEN>" \
     http://localhost:9000/api/tools/<id>/
```

Closes #634
